tree/ntuple/v7/inc/ROOT/RNTupleZip.hxx

```diff
@@ -0,0 +1,191 @@
+/// \file ROOT/RNTupleZip.hxx
+/// \ingroup NTuple ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \date 2019-11-21
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RNTupleZip
+#define ROOT7_RNTupleZip
+
+#include <RZip.h>
+#include <TError.h>
+
+#include <algorithm>
+#include <array>
+#include <cstring>
+#include <functional>
+#include <memory>
+#include <utility>
+
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RNTupleCompressor
+\ingroup NTuple
+\brief Helper class to compress data blocks in the ROOT compression frame format
+*/
+// clang-format on
+class RNTupleCompressor {
+private:
+   using Buffer_t = std::array<unsigned char, kMAXZIPBUF>;
+   std::unique_ptr<Buffer_t> fZipBuffer;
+
+public:
+   /// Data might be overwritten, if a zipped block in the middle of a large input data stream
+   /// turns out to be uncompressible
+   using Writer_t = std::function<void(const void *buffer, size_t nbytes, size_t offset)>;
+   static constexpr size_t kMaxSingleBlock = kMAXZIPBUF;
+
+   RNTupleCompressor() : fZipBuffer(std::unique_ptr<Buffer_t>(new Buffer_t())) {}
+   RNTupleCompressor(const RNTupleCompressor &other) = delete;
+   RNTupleCompressor &operator =(const RNTupleCompressor &other) = delete;
+
+   /// Returns the size of the compressed data. Data is compressed in 16MB blocks and written
+   /// piecewise using the provided writer
+   size_t operator() (const void *from, size_t nbytes, int compression, Writer_t fnWriter) {
```

Comment:
Shouldn't we use the newer compression interface rather than the encoded (100*a+l) int to selection the compression algorithm?

Reply:
I found the `int` more handy as a compact way to pass around the compression settings (otherwise I'd need two parameters, I think: algorithm and level).  This is in any case an internal interface.

Reply:
ok.